### PR TITLE
fix: Address memory leak that can occur if an instance loses all strong references before a scheduled repeating task begins.

### PR DESCRIPTION
**Pull request by scottmarchant**
### Motivation:

The primary motivation is to reduce bugs that may occur due to a memory leak in NIOCore. A new unit test demonstrates one known case where the memory leak could occur.

### Modifications:

Added `[weak self]` to a closure that begins repeating tasks.

### Result:

Before: ❌ Unit test fails. Memory leak occurs if the last remaining strong reference to a class or actor goes away before the repeating task begins.

After: ✅ Unit test succeeds. Memory does not leak and owning class successfully deinitializes.

### CODE REVIEW

File: Sources/NIOCore/EventLoop.swift [hunk 1]
```diff
@@ -90,8 +90,8 @@ public final class RepeatedTask {
         if self.eventLoop.inEventLoop {
             self.begin0(in: delay)
         } else {
-            self.eventLoop.execute {
-                self.begin0(in: delay)
+            self.eventLoop.execute { [weak self] in
```
Comment (scottmarchant) [inline Sources/NIOCore/EventLoop.swift:93]: Holding this change in Draft for a bit while I investigate potential side effects of this change. It seems existing code in the unit tests rely on this leak to ensure the task starts running, when those tests don't hold on to the returned `RepeatedTask` handle. That means this could break existing code usages. Investigating if there is a cleaner way to address this leak without causing that issue.

**Comment by scottmarchant** [comment]
Alright, based on my latest round of testing, it looks like this `[weak self]` change causes more issues than it fixes. I'm just going to close this for now.

**Comment by Lukasa** [comment]
@scottmarchant did you observe a leak due to this reference after an EventLoop was shutdown? Any circular references here should be broken by shutting down the loop, but it’s possible there is a bug there and if there is we should fix it.

**Comment by scottmarchant** [comment]
> @scottmarchant did you observe a leak due to this reference after an EventLoop was shutdown? Any circular references here should be broken by shutting down the loop, but it’s possible there is a bug there and if there is we should fix it.

@Lukasa From everything I've seen so far, shutting down the loop cleans everything up. I think there is an edge case here where things can get leaked if the repeating task never fires before the owner loses it's last reference - up until the shutdown event. The unit test here shows that edge case. This behavior of holding the owner up to the shutdown might be desired, though. But if the event loop has a very long lifetime, then that final strong reference could be an issue.

I think that edge case is likely to be very very small in practice, and the breakage caused by fixing it the way I tried to fix it in this diff could be quite problematic for people, since it changes behavior without changing API's. There might be a different way to address the problem that is safer, though.

I'm keeping an eye on this while working on some other nio-related efforts. If I make additional discoveries on this, I'll report back here with any findings.

**Comment by Lukasa** [comment]
Great, thanks for that @scottmarchant. As an FYI for any future fix: we'll want to avoid `weak` references at all costs due to their pervasive long-running performance costs. That goes particularly for any commonly-held objects like `EventLoop`s. So any future fix will need to be focused on making sure that explicit loop breaking actually does what it needs to do.

**Comment by scottmarchant** [comment]
> Great, thanks for that @scottmarchant. As an FYI for any future fix: we'll want to avoid `weak` references at all costs due to their pervasive long-running performance costs. That goes particularly for any commonly-held objects like `EventLoop`s. So any future fix will need to be focused on making sure that explicit loop breaking actually does what it needs to do.

@Lukasa Thanks, appreciate the guidance and additional insight!
